### PR TITLE
Search accounts & transactions

### PR DESCRIPTION
**Pull request by ilayaraja89**
Changelog:
- Add and update accounts with search `data`
- Update transactions with search `data`
- Search accounts and transactions with `id`, `terms` and `range`

### CODE REVIEW

**Comment by sudhirj** [comment]
I assume you're considering using a JSONB datatype for the tags? Let's not - limit to simple string key -> value for now, and use the separate tags table in the schema. 

The issue is that as the system grows larger, we'll start exporting data to Redshift / Bigquery and running read queries from there - and those systems will not support JSONB. 

So for this project stick to standard database structures as much as possible - can assume Postgres for writes, but we need to assume reads may happen from MySQL compatible engines.

**Comment by sudhirj** [comment]
Some operations are still possible with simple key values - equality, not equals, less than, greater than, convert to date, to integer etc. Should be possible to get a lot of the JSON functionality with a combination of these. Clients can also simulate nested keys by using their own dot or colon notation.

**Comment by ilayaraja89** [comment]
@sj Updated the Readme based on the comments.

**Comment by sudhirj** [comment]
How do we plan to store the tags if not on the separate tags tables?

**Comment by sudhirj** [comment]
Let's rephrase the tags APIs as 
'Transactions can have a JSON `data`

**Comment by sudhirj** [comment]
Can make it non-nullable with a default `{}`.

**Comment by sudhirj** [comment]
@ilayaraja89 Keep committing the ongoing work? Want to see how we're handling the query filters.

**Comment by ilayaraja89** [comment]
@sj Can you review this PR ?

File: controllers/accounts.go [hunk 1]
```diff
@@ -3,21 +3,122 @@ package controllers
...
-func GetAccountInfo(w http.ResponseWriter, r *http.Request, context *ledgerContext.AppContext) {
-	accountsDB := models.AccountDB{DB: context.DB}
+func GetAccounts(w http.ResponseWriter, r *http.Request, context *ledgerContext.AppContext) {
+	engine, err := models.NewSearchEngine(context.DB, "accounts")
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	defer r.Body.Close()
```
Comment (anuradhag) [inline controllers/accounts.go]: Can we move this before initialising NewSearchEngine? If it returns error we would not really use the engine object.

File: controllers/accounts.go [hunk 2]
```diff
@@ -3,21 +3,122 @@ package controllers
...
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
 
-	id := r.FormValue("id")
-	account := accountsDB.GetByID(id)
-	data, err := json.Marshal(account)
+	data, jerr := json.Marshal(results)
 	if err != nil {
```
Comment (anuradhag) [inline controllers/accounts.go]: This should be jerr

File: controllers/transactions.go [hunk 3]
```diff
@@ -64,3 +65,79 @@ func MakeTransaction(w http.ResponseWriter, r *http.Request, context *ledgerCont
...
 }
+
+func GetTransactions(w http.ResponseWriter, r *http.Request, context *ledgerContext.AppContext) {
+	engine, err := models.NewSearchEngine(context.DB, "transactions")
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	defer r.Body.Close()
```
Comment (anuradhag) [inline controllers/transactions.go]: Can we move this before initialising NewSearchEngine? If it returns error we would not really use the engine object.

**Comment by sudhirj** [comment]
Looks okay to me, but impossible to tell for sure without a variety of tests. Try doing something similar to the account balance CSV system, but use JSON to specify the search request and specify the expected list of ids to be returned.

**Comment by sudhirj** [comment]
Same as the transactions, let's post a bunch of accounts and transactions from a json, and then run a series of queries (specified in another JSON) and check if the results match.

**Comment by sudhirj** [comment]
Can do that in another PR, though. Let's get this out and get some feedback.

**Comment by ilayaraja89** [comment]
@anuradhag @sj The review comments are addressed. I will be working the Search test cases in a separate branch.

**Comment by ilayaraja89** [comment]
@sudhirj Do you have any changes in this PR ? If it is merged to master, we shall use this for Qbank.

**Comment by sudhirj** [comment]
@ilayaraja89 go ahead, we'll need to change it according to he requirements in qbank anyway.

**Pull request closed without merging**